qiskit_nature/hdf5.py

```diff
@@ -0,0 +1,160 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Qiskit Nature HDF5 Integration."""
+
+import importlib
+import logging
+import sys
+from pathlib import Path
+from typing import Any, Generator
+
+import h5py
+
+from qiskit_nature import QiskitNatureError
+
+if sys.version_info >= (3, 8):
+    # pylint: disable=no-name-in-module
+    from typing import runtime_checkable, Protocol
+else:
+    from typing_extensions import runtime_checkable, Protocol
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+@runtime_checkable
+class HDF5Storable(Protocol):
+    """A Protocol implemented by those classes which support conversion methods for HDF5."""
+
+    def to_hdf5(self, parent: h5py.Group) -> None:
+        """Stores this instance in a HDF5 group inside of the provided parent group.
+
+        Qiskit Nature uses the convention of storing the `__module__` and `__class__` information as
+        attributes of an HDF5 group. Furthermore, a `__version__` should be stored in order to allow
+        version handling at runtime.
```

Comment:
Yes the documentation needs to be improved here. Currently I am doing the following:
- the HDF5 attribute key is called `__version__` to reflect that this is for internal purposes only
- the value being stored in there is a class-variable called `VERSION`. This is class-dependent (starting at 1) which means that we can update the version of each class when it's serialization changes.

As per how to handle things:
Once a class gets its `VERSION` bumped, the `from_hdf5` method will need to be adapted to run a check against the HDF5 `__version__` attribute. If it is up-to-date it can proceed normally. If it is younger it should fallback to the version's correct loading behavior. This means that in the far future the `from_hdf5` methods may need to be able to handle multiple versions. However, I do not think that there will be that many version inceases (because now each object has it's own version).

Reply:
Makes sense, ideally it would be nice for some deprecation/migration period for older data. Maybe there will be some limits that it would refuse to handle it if too far out - maybe it should never attempt to handle data created by some newer version than you're running

Reply:
Yep those are reasonable suggestions. Should I state this explicitly in the docs? If so, we would probably need to formalize the length of the deprecation/migration period. Or should we simply remove support for legacy versions once the standard Qiskit-deprecation length has passed?

Reply:
With the former hdf5 we had support in there for reading the older version of the file and did not require them to be updated. With the parts here being more independent I am not sure - I guess it depends how things evolve if/when they do. Hopefully they can always read some older version where they would have to default, or otherwise deal with new parameter values that would be read from some newer version but did not exist back then. Perhaps guidelines are all that can be reasonably put here outlining what we'd like to see and whose responsibility it is over version. Does the outer framework check in anyway against the component or is all down to each component itself etc.

Reply:
Alright that makes sense :+1: I will update the docs to reflect these expectations.

It is all up to each component itself. I'm not sure if we can do anything useful from the outer framework here.